### ISSUE TITLE
Get-Item (Resolve-Path, and maybe others?) does not trim input, causing bad side-effects

### ISSUE THREAD

**Issue by jespergustinmsft** [PowerShell/PowerShell#16922]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

Noticed a dot (.) showing up in my filenames when getting the $item.BaseName property, and it turns out Get-Item and Resolve-Path doesn't trim the input from leading and trailing whitespaces.

In this minimized test case, I suffixed the path to Get-Item with 4 spaces, causing the BaseName property to return the complete file name, including the file extension:
```New-Item "foo.bar" | Out-Null; "`"$((Get-Item "foo.bar    ").BaseName)`""; Remove-Item "foo.bar"```

And this minimized test case throws an error that the file name "  foo.bar" cannot be found:
```New-Item "foo.bar" | Out-Null; "`"$((Get-Item "  foo.bar").BaseName)`""; Remove-Item "foo.bar"```

Resolve-Path behaves the same way, actually returning a path that ends with spaces:
```New-Item "foo.bar" | Out-Null; "`"$((Resolve-Path "foo.bar    ").Path)`""; Remove-Item "foo.bar"```

New-Item **does** trim trailing white-space however, so there is a discrepancy here on how input paths are handled.

### Expected behavior

```console
PS> New-Item "foo.bar" | Out-Null; "`"$((Get-Item "foo.bar    ").BaseName)`""; Remove-Item "foo.bar"
"foo"
```

### Actual behavior

```console
PS> New-Item "foo.bar" | Out-Null; "`"$((Get-Item "foo.bar    ").BaseName)`""; Remove-Item "foo.bar"
"foo.bar"
```

### Environment data

```powershell
PS> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.2.1
PSEdition                      Core
GitCommitId                    7.2.1
OS                             Microsoft Windows 10.0.22000
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by jhoneill** [comment]
Without quotes `(Get-item  foo.bar                  ).basename`  works as expected.   

Using quotes around `"foo bar . temp"`  makes the spaces part of the name.  It is valid (but unusual) to have leading spaces in a file name (trailing spaces is less usual but also valid) - or indeed in the extension. 
Specifying a trim copes with case that the user didn't type what they intended, but it's a sure bet that if `trim()` is applied to input someone will complain that they tried to do `set-content` or `out-file` and include spaces in the name and PowerShell won't let them. 
Isn't the answer to this (a) only use quotes on the command line where needed and (b) only around the characters you intend to be part of the file name ? 

Does any other shell do a it as suggested ?

**Comment by iSazonov** [comment]
Notice, Windows path API always trim trailing spaces.

**Comment by jespergustinmsft** [comment]
@jhoneill - one part of the problem is that BaseName property claims to return the file name without the file extension, so if trimming the input isn't the problem (or at least is behaving as expected), then the BaseName implementation needs to be fixed, since I would never expect anything but "foo" to be returned by that property.

As I wrote however - New-Item trims the input, which means you can never create any new files using New-Item that has trailing whitespace, which then adds another bug report to the stack if your idea that trailing whitespace should be valid(?). It does keep leading whitespace though and I have updated my report to reflect this.

**Comment by jhoneill** [comment]
> Notice, Windows path API always trim trailing spaces.

Yes, but not leading ones. 

`dir > "   temp.txt"`  Creates a file with leading spaces     
`dir \ > "   temp.txt  "`  overwrites it because trailing spaces are trimmed.

And searching for files 
```
ps >  dir "   temp.txt" | % {"***$($_.name)***"}
***   temp.txt***
ps > dir "   temp.txt              " | % {"***$($_.name)***"}
***   temp.txt              ***
```
I can put in any number of trailing spaces and that's the name that is returned even though the file name doesn't have them 

BaseName is weird  - it looks like a clunky bit of string parsing inside .NET (not PowerShell). "How many letters after the final . " , "Reduce the string length by that number." 
```
ps > dir " *" | % {"***$($_.BaseName)***"}
***   temp***
ps > dir "   temp.txt     " | % {"***$($_.BaseName)***"}
***   temp.txt ***
```
That does look a bug to take to the .NET folks.

Generally PowerShell doesn't change a path for you , if you put a : " > | < etc in the name or use a file name like LPT1 it finds its way into .NET and probably down into the OS, where the Windows filesystem (not .NET or PowerShell) generates the error and pushes it back up.

For other providers 
`dir "Cert:\CurrentUser\My\DB...23   " `  returns the cert   (trims the trailing space) 
`dir 'HKCU:\Control Panel ' ` returns an error (looks for the name with a trailing space ? )

Possibly the providers could impose some consistency, but I'm not convinced the effort to do it would be justified

**Comment by jhoneill** [comment]
It is legal to have trailing spaces in the *name* (base name) but not in the *extension*. What I wrote implies that the file-name-including-extension can have trailing spaces which isn't correct, sorry for creating confusion. 

1. Basename is implemented by .Net, does look like a bug and should be sent over there (perhaps by the PowerShell team).  
2. PowerShell providers appear to be lazy and pass unmodified names to whatever the providing for, I _think_ (for example) the registry API will create a key or value with spaces at the end, but the file system trims spaces from the end of the path (so after the extension).  Having the registry provider treat some legal paths as invalid to make all providers do the same thing is probably not a good path to go down
3. No, what @iSazonov  said above is correct: the windows API trims anything at the end of the path. I gave an example of spaces around the dot between name and extension, which works in Windows explorer , PowerShell etc. But if you try to add spaces at the very end, they're removed in PowerShell and Explorer.  I don't know what rules linux or mac file systems apply, and can't test from the machine I'm on, but I would say it is for the file system to decide how a given name is handled (if at all -the windows fs won't let you use "lpt1" but I think linux will, and I think there are differences over illegal characters, the provider shouldn't intervene). 
  
**TLDR**  Basename is bug for .NET , the rest is by design.

**Comment by iSazonov** [comment]
> **TLDR** Basename is bug for .NET , the rest is by design.

Basename is not .Net property, it comes from PowerShell ETS.

**Comment by jhoneill** [comment]
> > **TLDR** Basename is bug for .NET , the rest is by design.
> 
> Basename is not .Net property, it comes from PowerShell ETS.
> 
> https://github.com/PowerShell/PowerShell/blob/ee6bc9527c2596f048088fa7b396d2664a11ad08/src/System.Management.Automation/engine/TypeTable_Types_Ps1Xml.cs#L667
> 
> https://github.com/PowerShell/PowerShell/blob/ee6bc9527c2596f048088fa7b396d2664a11ad08/src/System.Management.Automation/engine/TypeTable_Types_Ps1Xml.cs#L803

Then that's my mistake. Should have looked harder - 

```
if ($this.Extension.Length -gt 0){$this.Name.Remove($this.Name.Length - $this.Extension.Length)}  else{$this.Name};
```
 
I'd rewrite that as ` $this.name -replace '\..*?$','' `

**Comment by jespergustinmsft** [comment]
Thank you for the input and help! :)

And just to be super clear - do we still think that:
```
PS> New-Item "foo.bar" | Out-Null; "`"$((Resolve-Path "foo.bar   ").Path)`""; Remove-Item "foo.bar";
"C:\test\foo.bar   "
```
is not a bug, and does not need to be fixed? To me that is still quite weird as well. IMO Resolve-Path should actually resolve and return the *actual* file name, and not be affected by how the input is given!?

**Comment by jhoneill** [comment]
A bit of double think here. I don't think it is strictly a bug, but I do agree that what we expect to come back is the actual path of the file. 

In this example we would not expect something which comes back from `Resolve-Path`  and is deemed OK by `Test-Path` to produce something invalid when used with `Join-Path`
```
 >$myDir = resolve-path  $SomeStringWeDoNotControl
> if ($myDir -and (Test-Path $myDir) ) {$myPsDir = Join-Path $mydir "PowerShell"} else {Write-warning "Bad Directory"}
> $mypsdir
C:\Users\James\Documents     \PowerShell
```

.Net can't decide what to do.  It trims trailing spaces from the full path, but keeps them from the name.  It should trim both or neither. 
```
 Get-item 'C:\Users\James\Documents\PowerShell\profile.ps1  ' | %{ "'$($_.fullname)'", "'$($_.name)'" }
'C:\Users\James\Documents\PowerShell\profile.ps1'
'profile.ps1  '
```

So `Get-Item` I would leave untouched but ask the .NET folks to be consistent, and I'd have **all** the `-path` commands trim their inputs **on the right only**  (I haven't checked if  `split-path` keeps or removes trailing spaces, but I'd guess it keeps them)

**Comment by daxian-dbw** [comment]
The engine WG reviewed this issue. It turns out to be an issue with `FileInfo` on Windows. Simply put, `FileInfo` somehow trims the path passed in when it should not. Then its `FullName` and `Extension` properties have the trailing spaces trimmed, but its `Name` property still has the trailing spaces kept. For details, see https://github.com/dotnet/runtime/issues/70524. Given that it's a .NET issue, we will resolve it as external.

**Comment by dprimon** [comment]
I report here a traling white space related issue I encountered today.

On a up-to-date Windows 10 Powershell 7.5.0 you may issue:

`PS> New-Item -ItemType Directory 'coco /pop'`

That creates the two nested directories, but `coco ` doesn't get trimmed ending in a stale folder that cannot be renamed/deleted at minimum. I still haven´t figured out out to remove that folder from the disk.
